pkg/driver/controller.go

```diff
@@ -243,6 +243,48 @@ func (cs *controller) init() error {
 	return nil
 }
 
+func validateRestoreRequestFromSnapshot(lvmSnapshotCrName string,
+	req *csi.CreateVolumeRequest,
+	params *VolumeParams,
+) error {
+	// get the source snapshot LVM custom resource
+	lvmSnapCr, err := lvm.GetLVMSnapshot(lvmSnapshotCrName)
+	if err != nil {
+		return status.Errorf(codes.Internal, "snapshot %s not found: %s", lvmSnapshotCrName, err.Error())
+	}
+
+	// verify that the snapshot is of thin type
+	// only thin snapshot can be used to restore a volume
+	// also verify that the new volume to be created from snapshot
+	// is also thin provisioned
+	if lvmSnapCr.Spec.ThinSnapshot != lvm.YES {
+		return status.Errorf(codes.Internal, "snapshot %s is not of thin type", lvmSnapshotCrName)
+	} else if params.ThinProvision != lvm.YES {
+		return status.Errorf(codes.Internal, "volume from snapshot %s must be thin provisioned", lvmSnapshotCrName)
+	}
+
+	vgPattern := fmt.Sprintf("^%v$", lvmSnapCr.Spec.VolGroup)
+
+	// verify that the volume group of the new volume to be created
+	// from snapshot is same as that of the snapshot
+	if params.VgPattern.String() != vgPattern {
+		return status.Errorf(codes.Internal, "volume group/ vg pattern of the new volume to be created from snapshot %s must be same as that of the snapshot",
+			lvmSnapshotCrName)
+	}
+
+	reqCapacity := strconv.FormatInt(getRoundedCapacity(
+		req.GetCapacityRange().RequiredBytes), 10)
+
+	// verify that the capacity of the new volume to be created
+	// from snapshot should equal to the source volume capacity at the time of snapshot creation
```

Comment:
This might need to change in future where we could automatically cater to larger sizes by doing extend at the time of restore.

```suggestion
	// from snapshot should equal to the origin volume capacity at the time of snapshot creation
```

Reply:
yes, it might change when larger restore volume capacity is supported.